src/peft/mapping.py

```diff
@@ -89,10 +99,37 @@ def get_peft_model(model: PreTrainedModel, peft_config: PeftConfig, adapter_name
 
     peft_config.base_model_name_or_path = model.__dict__.get("name_or_path", None)
 
-    if peft_config.task_type not in MODEL_TYPE_TO_PEFT_MODEL_MAPPING.keys() and not isinstance(
-        peft_config, PromptLearningConfig
-    ):
+    if peft_config.task_type not in MODEL_TYPE_TO_PEFT_MODEL_MAPPING.keys() and not peft_config.is_prompt_learning:
         return PeftModel(model, peft_config, adapter_name=adapter_name)
-    if isinstance(peft_config, PromptLearningConfig):
+    if peft_config.is_prompt_learning:
         peft_config = _prepare_prompt_learning_config(peft_config, model_config)
     return MODEL_TYPE_TO_PEFT_MODEL_MAPPING[peft_config.task_type](model, peft_config, adapter_name=adapter_name)
+
+
+def inject_adapter_in_model(peft_config: PeftConfig, model: torch.nn.Module, adapter_name: str):
+    r"""
+    A simple API to create and inject adapter in-place into a model. Currently the API does not support prompt learning
+    methods and adaption prompt. Make sure to have the correct `target_names` set in the `peft_config` object. The API
+    calls `get_peft_model` under the hood but would be restricted only to non-prompt learning methods.
+
+    Args:
+        peft_config (`PeftConfig`):
+            Configuration object containing the parameters of the Peft model.
+        model (`torch.nn.Module`):
+            The input model where the adapter will be injected.
+        adapter_name (`str`):
+            The name of the adapter to be injected.
+    """
+    if peft_config.is_prompt_learning or peft_config.is_adaption_prompt:
+        raise ValueError("`create_and_replace` does not support prompt learning and adaption prompt yet.")
+
+    if peft_config.peft_type not in PEFT_TYPE_TO_TUNER_MAPPING.keys():
+        raise ValueError(
+            f"`inject_adapter_in_model` does not support {peft_config.peft_type} yet. Please use `get_peft_model`."
+        )
+
+    tuner_cls = PEFT_TYPE_TO_TUNER_MAPPING[peft_config.peft_type]
+
+    peft_model = tuner_cls(model, peft_config, adapter_name=adapter_name)
```

Comment:
If I went down the rabbit hole correctly [here](https://github.com/huggingface/peft/blob/ec267c644a9a9f05a7340a7cb23ed5a6a6090dd0/src/peft/tuners/lora.py#L202) it looks like this line will set `required_grad=False` for all weights of the model. For https://github.com/huggingface/transformers/pull/25077 this would mean that if a load lora weights of a model all weights of the model (except the lora weights) are frozen - I would not expect such a behavior here.

Also it seems like the peft config decides further down the road here: https://github.com/huggingface/peft/blob/ec267c644a9a9f05a7340a7cb23ed5a6a6090dd0/src/peft/tuners/lora.py#L203 if we're in inference mode or not - for the transformers implementation I would make try to not make any assumptions here depending on the config but leave it fully up to the user - so just not do anything (don't freeze any grads). 

Can we add a flag here that would allow for this ?

Also I'd maybe add a comment to make it a bit easier for the user to understand what this line does.
```suggestion
    # By instantiating a peft model we are injecting randomly initialized LoRA layers into the model's modules.
    peft_model = tuner_cls(model, peft_config, adapter_name=adapter_name)
```

Reply:
Freezing the base model should be optional when you load a model I think you are right

However when you attach fresh new adapters usually (for 99% of the usecases) it is to train them, so maybe we should make a distinction

1- load a pretrained adapter --> not necessearly freeze the base model
2- add a new random adapter --> I would expect users to train the adapters --> freeze the base model

WDYT? @pacman100 @BenjaminBossan

Reply:
That makes sense, I would maybe give the user full flexibility here then and add a new function argument:
`freeze_weights=True/False/"non-lora-only"` or something similar? I think when calling this from `transformers` we don't want all the model to be frozen, but for other use cases it could very much be the case I'd say

Reply:
> 1- load a pretrained adapter --> not necessearly freeze the base model
> 2- add a new random adapter --> I would expect users to train the adapters --> freeze the base model

> That makes sense, I would maybe give the user full flexibility here then and add a new function argument

Personally, I would prefer to give the user an explicit argument, rather than (or in addition to) trying to guess based on the circumstances, so I would go with that option.

@patrickvonplaten: You bring up some good points. I will still merge the PR as is because Younes will be OoO for some time and we don't want this big refactor to become stale. We should be able to address your concerns in a follow up PR.

Reply:
Sounds good! Agree that giving the user an explicit argument is the right option here